File: java/methods/BruteForce.java
/*
Resolução de equação não-linear transcendental: método da força bruta.
Programa escrito em java que implementa os métodos da classe BruteForce.
Autor: Ubiratan Tavares
Data: 07 de novembro de 2024.
*/

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class BruteForce {
    
    private Function<Double, Double> func;
    private double a;
    private double b;
    private int numPoints;
    private List<double[]> roots;

    // Construtor para inicializar a função, os limites e o número de pontos
    public BruteForce(Function<Double, Double> func, double a, double b, int numPoints) {
        this.func = func;
        this.a = a;
        this.b = b;
        this.numPoints = numPoints;
        this.roots = new ArrayList<>();
    }

    // Método para calcular o passo
    private double calculateStep() {
        return (b - a) / numPoints;
    }

    // Método para verificar se os sinais são opostos
    private boolean areOppositeSigns(double x1, double x2) {
        return func.apply(x1) * func.apply(x2) < 0;
    }

    // Método para encontrar raízes usando força bruta
    public List<double[]> findRoots() {
        double step = calculateStep();
        roots.clear();

        for (int i = 0; i <= numPoints; i++) {
            double x = a + i * step;
            if (areOppositeSigns(x, x + step)) {
                roots.add(new double[] {x, x + step});
            }
        }
        return roots;
    }

    // Método para verificar se a lista está vazia
    private boolean isListEmpty() {
        return roots.isEmpty();
    }

    // Método para exibir o resultado
    public void displayResult() {
        if (isListEmpty()) {
            System.out.println("Não encontrou raíz(es) no intervalo.");
        } else {
            System.out.println("Há raíz(es) no(s) sub-intervalo(s):");
            for (double[] interval : roots) {
                System.out.printf("(%f, %f) ", interval[0], interval[1]);
            }
            System.out.println();
        }
    }
}

File: c/methods/brute_force.c
/*
Resolução de equação algébrica pelo método de força bruta.
Autor: Ubiratan Tavares
Data: 07 de novembro de 2024.
Comando para compilação: gcc -o brute_force brute_force.c -lm
*/

#include <stdio.h>
#include <math.h>

// função para calcular o passo
double calculate_step(double a, double b, int num_points) {
    return (b - a) / num_points;
}

// função para verificar se os sinais são opostos
int are_opposite_signs(double (*func)(double), double a, double b) {
    return func(a) * func(b) < 0;
}

// função de busca forçada para encontrar raízes
void brute_force(double (*func)(double), double a, double b, int num_points) {
    double step = calculate_step(a, b, num_points);
    int found_root = 0;
    printf("Intervalos onde há raízes\n");
    for (int i = 0; i <= num_points; i++) {
        double x = a + i * step;
        if (are_opposite_signs(func, x, x + step)) {
            printf("(%lf, %lf)\n", x, x + step);
            found_root = 1;
	}
    }
    if (!found_root) {
        printf("Não encontrou raíz(es) no intervalo.\n");
    }
}

// função que define f(x) = e^x - 2
double function(double x) {
    return  exp(x) - 2;
}

// função principal
int main() {
    double a =  0.0, b = 1.0;
    int num_points = 100000;
    brute_force(function, a, b, num_points);
    return 0;
}
